lib/setup/setupUpgrade.js

```diff
@@ -50,29 +51,116 @@ function Upgrade(options) {
      */
-    this.upgradeAdapterHelper = function (repoUrl, list, i, forceDowngrade, autoConfirm, callback) {
+    this.upgradeAdapterHelper = (repo, list, forceDowngrade, autoConfirm, callback) => {
         if (typeof autoConfirm === 'function') {
             callback = autoConfirm;
             autoConfirm = false;
         }
 
-        this.upgradeAdapter(repoUrl, list[i], forceDowngrade, autoConfirm, true, () => {
+        const relevantAdapters = [];
+        // check which adapters are upgradeable and sort them according to their dependencies
+        for (const adapter of list) {
+            if (repo[adapter].controller) {
+                // skip controller
+                continue;
+            }
+            const adapterDir = tools.getAdapterDir(adapter);
+            if (fs.existsSync(`${adapterDir}/io-package.json`)) {
+                const ioInstalled = require(`${adapterDir}/io-package.json`);
+                if (!tools.upToDate(repo[adapter].version, ioInstalled.common.version)) {
+                    // not up to date, we need to put it into account for our dependency check
+                    relevantAdapters.push(adapter);
+                }
+            }
+        }
+
+        if (relevantAdapters) {
+            const sortedAdapters = [];
+
+            while (relevantAdapters.length) {
+                let oneAdapterAdded = false;
+                // create ordered list for upgrades
+                for (let i = relevantAdapters.length - 1; i >= 0; i--) {
+                    const relAdapter = relevantAdapters[i];
+                    // if new version has no dependencies we can upgrade
+                    if (!repo[relAdapter].dependencies && !repo[relAdapter].globalDependencies) {
+                        // no deps, simply add it
+                        sortedAdapters.push(relAdapter);
+                    } else {
+                        const deps = repo[relAdapter].dependencies || [];
+                        const globalDeps = repo[relAdapter].globalDependencies || [];
+
+                        // we have to check if the deps are there
+                        let conflict = false;
+                        for (const dep of [...deps, ...globalDeps]) {
+                            debug(`adapter "${relAdapter}" has dependency "${JSON.stringify(dep)}"`);
+                            if (tools.isObject(dep)) {
+                                // object -> dependency is important, because it affects version range
+                                for (const depName of Object.keys(dep)) {
+                                    if (relevantAdapters.includes(depName)) {
+                                        // the dependency is also in the upgrade list and not previously added, we should add the dependency first
+                                        debug(`conflict for dependency "${depName}" at adapter "${relAdapter}"`);
+                                        conflict = true;
```

Comment:
Maybe still at them at the very end to "hope" that it works?! else will give errr then on upgrade or!?

Reply:
Conflict means, I am an adapter and I depend on other adapters which are not yet in the upgrade list, so they should be in upgrade list before I can be added. _There is an outer loop which will add me later, when the other adapters which I depend on are added. Or what exactly do you mean?

Reply:
ahh ... ok. makes sense now